File: app/src/main/java/com/nononsenseapps/feeder/util/PrefUtils.java
package com.nononsenseapps.feeder.util;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;

/**
 * Utilities and constants related to app preferences.
 */
public class PrefUtils {
    /**
     * Boolean indicating whether we performed the (one-time) welcome flow.
     */
    public static final String PREF_WELCOME_DONE = "pref_welcome_done";

    /**
     * Boolean indicating if night mode should be engaged.
     */
    public static final String PREF_NIGHT_MODE = "pref_night_mode";

    /**
     * Boolean indicating if only unread items should be shown
     */
    public static final String PREF_SHOW_ONLY_UNREAD = "pref_show_only_unread";

    /**
     * These indicate which fragment to open by default
     */
    public static final String PREF_LAST_FEED_TAG = "pref_last_feed_tag";
    public static final String PREF_LAST_FEED_ID = "pref_last_feed_id";

    /**
     * Sync settings
     */
    public static final String PREF_SYNC_ONLY_CHARGING = "pref_sync_only_charging";
    public static final String PREF_SYNC_HOTSPOTS = "pref_sync_hotspots";
    public static final String PREF_SYNC_ONLY_WIFI = "pref_sync_only_wifi";
    public static final String PREF_SYNC_FREQ = "pref_sync_freq";

    /**
     * Image settings
     */
    public static final String PREF_IMG_ONLY_WIFI = "pref_img_only_wifi";
    public static final String PREF_IMG_HOTSPOTS = "pref_img_hotspots";

    public static boolean shouldLoadImagesOnlyOnWIfi(final Context context) {
        return sp(context).getBoolean(PREF_IMG_ONLY_WIFI, false);
    }

    public static boolean shouldLoadImagesOnHotSpots(final Context context) {
        return sp(context).getBoolean(PREF_IMG_HOTSPOTS, false);
    }

    public static boolean shouldSyncOnlyOnWIfi(final Context context) {
        return sp(context).getBoolean(PREF_SYNC_ONLY_WIFI, false);
    }

    public static boolean shouldSyncOnlyWhenCharging(final Context context) {
        return sp(context).getBoolean(PREF_SYNC_ONLY_CHARGING, false);
    }

    public static boolean shouldSyncOnHotSpots(final Context context) {
        return sp(context).getBoolean(PREF_SYNC_HOTSPOTS, false);
    }

    /**
     * @return true if automatic syncing is enabled, false otherwise
     */
    public static boolean shouldSync(final Context context) {
        return synchronizationFrequency(context) > 0;
    }

    /**
     * @return number of minutes between syncs, or zero if none should be performed
     */
    public static long synchronizationFrequency(final Context context) {
        return Long.parseLong(sp(context).getString(PREF_SYNC_FREQ, "60"));
    }

    /**
     * A shorthand method
     */
    private static SharedPreferences sp(final Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    public static boolean isWelcomeDone(final Context context) {
        return sp(context).getBoolean(PREF_WELCOME_DONE, false);
    }

    public static void markWelcomeDone(final Context context) {
        sp(context).edit().putBoolean(PREF_WELCOME_DONE, true).apply();
    }

    public static boolean isNightMode(final Context context) {
        return sp(context).getBoolean(PREF_NIGHT_MODE, false);
    }

    public static void setNightMode(final Context context, final boolean value) {
        sp(context).edit().putBoolean(PREF_NIGHT_MODE, value).apply();
    }

    public static boolean isShowOnlyUnread(final Context context) {
        return sp(context).getBoolean(PREF_SHOW_ONLY_UNREAD, true);
    }

    public static void setPrefShowOnlyUnread(final Context context,
                                             final boolean value) {
        sp(context).edit().putBoolean(PREF_SHOW_ONLY_UNREAD, value).apply();
    }

    /**
     * Remember the open fragment, either by id, or tag if id < 1.
     * Clears the other one.
     *
     * @param context
     * @param id
     * @param tag
     */
    public static void setLastOpenFeed(final Context context, final long id,
                                       final String tag) {
        sp(context).edit().putLong(PREF_LAST_FEED_ID, id)
                .putString(PREF_LAST_FEED_TAG, tag).apply();
    }

    /**
     * Get which feed tag was last open. Check id first.
     *
     * @param context
     * @return last open tag, or null
     */
    public static String getLastOpenFeedTag(final Context context) {
        return sp(context).getString(PREF_LAST_FEED_TAG, null);
    }

    /**
     * Get which feed id was last open. If -1, check tag.
     *
     * @param context
     * @return last open id, or -1
     */
    public static long getLastOpenFeedId(final Context context) {
        return sp(context).getLong(PREF_LAST_FEED_ID, -1);
    }


    public static void registerOnSharedPreferenceChangeListener(final Context context,
                                                                SharedPreferences.OnSharedPreferenceChangeListener listener) {
        sp(context).registerOnSharedPreferenceChangeListener(listener);
    }

    public static void unregisterOnSharedPreferenceChangeListener(final Context context,
                                                                  SharedPreferences.OnSharedPreferenceChangeListener listener) {
        sp(context).unregisterOnSharedPreferenceChangeListener(listener);
    }

    public static boolean shouldLoadImages(Context context) {
        if (SystemUtils.currentlyOnWifi(context)) {
            return !SystemUtils.currentlyMetered(context) || shouldLoadImagesOnHotSpots(context);
        }
        return !shouldLoadImagesOnlyOnWIfi(context);
    }

    /**
     * A preference value change listener that updates the preference's summary
     * to reflect its new value. Handles ringtone preferences and list preferences
     * specially.
     */
    public static final Preference.OnPreferenceChangeListener summaryUpdater = new Preference.OnPreferenceChangeListener() {
        @Override
        public boolean onPreferenceChange(@NonNull Preference preference, @NonNull Object value) {
            String stringValue = value.toString();
            if (preference instanceof ListPreference) {
                // For list preferences, look up the correct display value in
                // the preference's 'entries' list.
                ListPreference listPreference = (ListPreference) preference;
                int index = listPreference.findIndexOfValue(stringValue);
                // Set the summary to reflect the new value.
                preference.setSummary(index >= 0 ? listPreference.getEntries()[index] : null);
            } else {
                // For all other preferences, set the summary to the value's
                // simple string representation.
                preference.setSummary(stringValue);
            }
            return true;
        }
    };
}

File: app/src/main/java/com/nononsenseapps/feeder/util/PrefUtils.kt
package com.nononsenseapps.feeder.util

import android.app.Application
import android.content.SharedPreferences
import com.nononsenseapps.feeder.R
import com.nononsenseapps.feeder.archmodel.ItemOpener
import com.nononsenseapps.feeder.archmodel.LinkOpener
import com.nononsenseapps.feeder.archmodel.PREF_DEFAULT_OPEN_ITEM_WITH
import com.nononsenseapps.feeder.archmodel.PREF_IMG_ONLY_WIFI
import com.nononsenseapps.feeder.archmodel.PREF_IMG_SHOW_THUMBNAILS
import com.nononsenseapps.feeder.archmodel.PREF_LAST_FEED_ID
import com.nononsenseapps.feeder.archmodel.PREF_LAST_FEED_TAG
import com.nononsenseapps.feeder.archmodel.PREF_OPEN_LINKS_WITH
import com.nononsenseapps.feeder.archmodel.PREF_PRELOAD_CUSTOM_TAB
import com.nononsenseapps.feeder.archmodel.PREF_SHOW_FAB
import com.nononsenseapps.feeder.archmodel.PREF_SHOW_ONLY_UNREAD
import com.nononsenseapps.feeder.archmodel.PREF_SORT
import com.nononsenseapps.feeder.archmodel.PREF_SYNC_FREQ
import com.nononsenseapps.feeder.archmodel.PREF_SYNC_ONLY_CHARGING
import com.nononsenseapps.feeder.archmodel.PREF_SYNC_ONLY_WIFI
import com.nononsenseapps.feeder.archmodel.PREF_SYNC_ON_RESUME
import com.nononsenseapps.feeder.archmodel.PREF_VAL_OPEN_WITH_BROWSER
import com.nononsenseapps.feeder.archmodel.PREF_VAL_OPEN_WITH_CUSTOM_TAB
import com.nononsenseapps.feeder.archmodel.PREF_VAL_OPEN_WITH_READER
import com.nononsenseapps.feeder.archmodel.PREF_VAL_OPEN_WITH_WEBVIEW
import com.nononsenseapps.feeder.archmodel.PREF_WELCOME_DONE
import com.nononsenseapps.feeder.archmodel.SortingOptions
import com.nononsenseapps.feeder.archmodel.SyncFrequency
import com.nononsenseapps.feeder.db.room.ID_UNSET
import org.kodein.di.DI
import org.kodein.di.DIAware
import org.kodein.di.instance


/**
 * Database settings
 */
const val PREF_MAX_ITEM_COUNT_PER_FEED = "pref_max_item_count_per_feed"

fun SharedPreferences.getStringNonNull(key: String, defaultValue: String): String =
    getString(key, defaultValue) ?: defaultValue

// TODO delete me
class Prefs(override val di: DI) : DIAware {
    private val sp: SharedPreferences by instance()
    private val app: Application by instance()

    var onlyLoadImagesOnWIfi: Boolean
        get() = sp.getBoolean(PREF_IMG_ONLY_WIFI, false)
        set(value) = sp.edit().putBoolean(PREF_IMG_ONLY_WIFI, value).apply()

    var showThumbnails: Boolean
        get() = sp.getBoolean(PREF_IMG_SHOW_THUMBNAILS, true)
        set(value) = sp.edit().putBoolean(PREF_IMG_SHOW_THUMBNAILS, value).apply()

    var onlySyncOnWifi: Boolean
        get() = sp.getBoolean(PREF_SYNC_ONLY_WIFI, false)
        set(value) = sp.edit().putBoolean(PREF_SYNC_ONLY_WIFI, value).apply()

    var syncOnResume: Boolean
        get() = sp.getBoolean(PREF_SYNC_ON_RESUME, false)
        set(value) = sp.edit().putBoolean(PREF_SYNC_ON_RESUME, value).apply()

    var onlySyncWhileCharging: Boolean
        get() = sp.getBoolean(PREF_SYNC_ONLY_CHARGING, false)
        set(value) = sp.edit().putBoolean(PREF_SYNC_ONLY_CHARGING, value).apply()

    var maximumCountPerFeed: Int
        get() = sp.getStringNonNull(PREF_MAX_ITEM_COUNT_PER_FEED, "100").toInt()
        set(value) = sp.edit().putString(PREF_MAX_ITEM_COUNT_PER_FEED, "$value").apply()

    /**
     * @return true if automatic syncing is enabled, false otherwise
     */
    fun shouldSync(): Boolean = synchronizationFrequency > 0

    /**
     * @return number of minutes between syncs, or zero if none should be performed
     */
    var synchronizationFrequency: Long
        get() = sp.getStringNonNull(PREF_SYNC_FREQ, "60").toLong()
        set(value) = sp.edit().putString(PREF_SYNC_FREQ, "$value").apply()

    var openItemsWith: String
        get() = sp.getStringNonNull(PREF_DEFAULT_OPEN_ITEM_WITH, PREF_VAL_OPEN_WITH_READER)
        set(value) = sp.edit().putString(PREF_DEFAULT_OPEN_ITEM_WITH, value).apply()

    var openLinksWith: String
        get() = sp.getStringNonNull(PREF_OPEN_LINKS_WITH, PREF_VAL_OPEN_WITH_CUSTOM_TAB)
        set(value) = sp.edit().putString(PREF_OPEN_LINKS_WITH, value).apply()

    var preloadCustomTab: Boolean
        get() = sp.getBoolean(PREF_PRELOAD_CUSTOM_TAB, false)
        set(value) = sp.edit().putBoolean(PREF_PRELOAD_CUSTOM_TAB, value).apply()

    var welcomeDone: Boolean
        get() = sp.getBoolean(PREF_WELCOME_DONE, false)
        set(value) = sp.edit().putBoolean(PREF_WELCOME_DONE, value).apply()

    var currentItemOpener: ItemOpener
        get() = when (openItemsWith) {
            PREF_VAL_OPEN_WITH_BROWSER -> ItemOpener.DEFAULT_BROWSER
            PREF_VAL_OPEN_WITH_WEBVIEW,
            PREF_VAL_OPEN_WITH_CUSTOM_TAB,
            -> ItemOpener.CUSTOM_TAB
            else -> ItemOpener.READER
        }
        set(value) {
            openItemsWith = when (value) {
                ItemOpener.READER -> PREF_VAL_OPEN_WITH_READER
                ItemOpener.CUSTOM_TAB -> PREF_VAL_OPEN_WITH_CUSTOM_TAB
                ItemOpener.DEFAULT_BROWSER -> PREF_VAL_OPEN_WITH_BROWSER
            }
        }

    var currentLinkOpener: LinkOpener
        get() = when (openLinksWith) {
            PREF_VAL_OPEN_WITH_BROWSER -> LinkOpener.DEFAULT_BROWSER
            else -> LinkOpener.CUSTOM_TAB
        }
        set(value) {
            when (value) {
                LinkOpener.CUSTOM_TAB -> openLinksWith = PREF_VAL_OPEN_WITH_CUSTOM_TAB
                LinkOpener.DEFAULT_BROWSER -> openLinksWith = PREF_VAL_OPEN_WITH_BROWSER
            }
        }

    var currentSyncFrequency: SyncFrequency
        get() =
            SyncFrequency.values()
                .firstOrNull {
                    it.minutes==synchronizationFrequency
                }
                ?: SyncFrequency.MANUAL
        set(value) {
            synchronizationFrequency = value.minutes
        }

    var currentSorting: SortingOptions
        get() = when (sp.getString(PREF_SORT, app.getString(R.string.pref_sort_value_default))) {
            app.getString(R.string.pref_sort_value_oldest_first) -> SortingOptions.OLDEST_FIRST
            else -> SortingOptions.NEWEST_FIRST
        }
        set(value) = sp.edit().putString(
            PREF_SORT,
            when (value) {
                SortingOptions.NEWEST_FIRST -> app.getString(R.string.pref_sort_value_newest_first)
                SortingOptions.OLDEST_FIRST -> app.getString(R.string.pref_sort_value_oldest_first)
            }
        ).apply()

    var isNewestFirst: Boolean
        get() = when (currentSorting) {
            SortingOptions.NEWEST_FIRST -> true
            else -> false
        }
        set(value) {
            currentSorting = when (value) {
                true -> SortingOptions.NEWEST_FIRST
                false -> SortingOptions.OLDEST_FIRST
            }
        }

    var show_fab: Boolean
        get() = sp.getBoolean(PREF_SHOW_FAB, true)
        set(value) = sp.edit().putBoolean(PREF_SHOW_FAB, value).apply()

    var showOnlyUnread: Boolean
        get() = sp.getBoolean(PREF_SHOW_ONLY_UNREAD, true)
        set(value) = sp.edit().putBoolean(PREF_SHOW_ONLY_UNREAD, value).apply()

    /**
     * Remember the open fragment, either by id, or tag if id < 1.
     * Clears the other one.
     *
     * @param id
     * @param tag
     */
    fun setLastOpenFeed(id: Long, tag: String?) =
        sp.edit().putLong(PREF_LAST_FEED_ID, id)
            .putString(PREF_LAST_FEED_TAG, tag)
            .apply()

    /**
     * Get which feed tag was last open. Check id first.
     */
    var lastOpenFeedTag: String?
        get() = sp.getString(PREF_LAST_FEED_TAG, null)
        set(value) = sp.edit().putString(PREF_LAST_FEED_TAG, value).apply()

    /**
     * Get which feed id was last open. If -1, check tag.
     */
    var lastOpenFeedId: Long
        get() = sp.getLong(PREF_LAST_FEED_ID, ID_UNSET)
        set(value) = sp.edit().putLong(PREF_LAST_FEED_ID, value).apply()

    fun shouldLoadImages(): Boolean {
        return when {
            onlyLoadImagesOnWIfi -> currentlyUnmetered(app)
            else -> true
        }
    }

    val shouldPreloadCustomTab: Boolean
        get() = preloadCustomTab && openLinksWith==PREF_VAL_OPEN_WITH_CUSTOM_TAB
}
